encoders/firebase-decoders-json/src/main/java/com/google/firebase/decoders/FieldRef.java

```diff
@@ -0,0 +1,77 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.decoders;
+
+import androidx.annotation.NonNull;
+
+// TODO: support required, optional.
+
+/**
+ * {@link FieldRef} act as an wire to put decoded information to {@link ObjectDecoderContext} and
+ * get decoded information from {@link CreationContext}. It also provide {@link DataDecoder} the
+ * ability to obtain the data type of nextToken.
+ */
+public abstract class FieldRef<T> {
+  private TypeToken<T> typeToken;
+
+  @NonNull
+  public abstract TypeToken<T> getTypeToken();
+
+  /** Used to represent primitive data type. */
+  public static final class Primitive<T> extends FieldRef<T> {
+    @NonNull public static Primitive<Integer> INT = new Primitive<>(TypeToken.of(int.class));
```

Comment:
Additionally, I'd suggest you move these constants out into `FieldRef` itself, and also move `Boxed#of()` into FieldRef as sell. It will make the public interface to `FieldRefs` less hierarchical which might make it simpler, i.e.:

```java
class FieldRef<T> {
  static Primitive<Boolean> BOOLEAN = //...;
  static Primitive<Short> SHORT = //...;
  // etc,
  static Boxed<T> of(TypeToken<T> token) {}
}
```